File: packages/shared/src/evidence.ts
import { z } from 'zod';
import { BoundingBox, VehicleClass } from './sighting.js';

/**
 * One best-shot crop on its way from the analytics worker to the object store.
 *
 * Its own stream and its own type rather than more fields on `Sighting`, for a reason that is about
 * rates and not about tidiness: `sightings` carries roughly one entry per detection per inferred
 * frame, and `evidence` carries one per *track session*, each about 20 KB. One bounded stream
 * cannot be trimmed correctly for both.
 *
 * The record identifies its sighting by `(cameraId, trackId, framePtsMs)`. That is unique **because
 * `trackId` is session-qualified** (`session * 100_000 + tracker_id`, D1-09): the raw ByteTrack id
 * is reused across a loop-point scene cut, and joining on it would attach one vehicle's crop to a
 * different vehicle's row.
 */
export const EvidenceRecord = z.object({
  /** The camera's EXTERNAL id (`cam01`), exactly as the `sightings` stream carries it. */
  cameraId: z.string().min(1),
  trackId: z.number().int().nonnegative(),
  /** PTS-derived wall clock of the source frame. Never the time the crop was produced. */
  ts: z.iso.datetime(),
  framePtsMs: z.number().nonnegative(),

  kind: z.enum(['vehicle', 'plate']).default('vehicle'),
  class: VehicleClass,
  detConfidence: z.number().min(0).max(1),
  bbox: BoundingBox,

  /** 0-1. How good this observation was as evidence, versus the rest of its track session. */
  bestShotScore: z.number().min(0).max(1),
  /** Variance of the Laplacian of the crop. Comparable within a camera, not across cameras. */
  focus: z.number().nonnegative(),
  /** Observations of this track that were considered before this one won. The compression ratio. */
  observations: z.number().int().positive().default(1),

  vehicleType: z.string().nullable().default(null),
  /** `unknown` when the classifier refused the read — never the runner-up quietly promoted. */
  vehicleColor: z.string(),
  vehicleColorConfidence: z.number().min(0).max(1),
  attributesLowConfidence: z.boolean(),
  /** Share of the voting region that carried any colour at all. ~0 on a night frame. */
  colorChromaShare: z.number().min(0).max(1).default(0),
  colorRunnerUp: z.string().nullable().default(null),

  /**
   * Vehicle appearance descriptor (D3-03), or `null` when none could be made.
   *
   * **Not biometric.** It describes the outside of a vehicle — white-balanced colour histograms over
   * four stripes of the crop plus a coarse edge signature. SAAKSHI performs no face recognition and
   * stores no biometric template; `docs/reid.md` §2 and migration `0022` both say so, because a
   * reader who sees "embedding" and assumes "face" would be wrong about the most sensitive thing in
   * the system.
   *
   * Optional and nullable so that a record produced before D3-03 — or by a worker whose embedder
   * failed — still validates. The consumer writes `sighting_appearance` only when both fields are
   * present, and two descriptors may only ever be compared when their `appearanceEmbedderId`s match.
   */
  appearanceEmbedderId: z.string().min(1).nullable().default(null),
  appearance: z.array(z.number()).min(1).max(4096).nullable().default(null),

  contentType: z.string().default('image/jpeg'),
  /** The JPEG. Base64 because both Valkey clients round-trip text safely and neither does bytes. */
  cropBase64: z.string().min(1),
  cropBytes: z.number().int().nonnegative(),
});
export type EvidenceRecord = z.infer<typeof EvidenceRecord>;

File: workers/analytics/evidence.py
"""The `evidence` Valkey stream — best-shot crops on their way to the object store.

A **second** stream, deliberately, rather than more fields on `sightings`:

* `sightings` carries ~1 entry per detection per inferred frame and must stay small and fast;
  `evidence` carries ~1 entry per *track session* and each one is a JPEG. Mixing a 200-byte
  high-rate stream with a 20 KB low-rate one gives the bounded stream one trim policy that is wrong
  for both.
* The MinIO credentials live in exactly one place — the API. The worker never holds them, never
  signs a request, and cannot write to the bucket even if it is compromised.
* D2-01 is extending the `sightings` payload with plate reads in the same milestone. Two tickets
  editing one wire format in one wave is a merge conflict with a data-loss failure mode.

Wire contract — D2-06 (alerts) and D3-04 (export bundles) both read what this produces:

    stream key   evidence
    entry        XADD evidence MAXLEN ~ 2000 * payload <json>
    payload      one EvidenceRecord, camelCase, `cameraId` carrying the camera's EXTERNAL id
                 ("cam01") exactly as the `sightings` stream does; `cropBase64` is the JPEG
    group        evidence-writer     (created MKSTREAM by packages/api/src/consumers/evidence.ts)

`MAXLEN ~ 2000` rather than the sightings stream's 200,000: at ~20 KB per entry that is a ~40 MB
ceiling on the broker, and a stopped consumer must not be able to take Valkey down. Best shots
arrive at roughly one per vehicle appearance, so 2,000 is over an hour of the eight-camera estate.

The consumer resolves the record to its `sightings` row by `(camera_id, track_id, frame_pts_ms)` —
which is unique *because* `track_id` is session-qualified. Joining on the raw tracker id would match
two different vehicles across a loop-point cut.
"""

from __future__ import annotations

import base64
import json
import logging
import os
from dataclasses import dataclass
from typing import Protocol

log = logging.getLogger("saakshi.analytics")

STREAM_KEY = "evidence"
CONSUMER_GROUP = "evidence-writer"
#: ~2,000 entries x ~20 KB ≈ 40 MB ceiling on the broker. Bounded on purpose; see the module note.
STREAM_MAXLEN = 2_000
DEFAULT_VALKEY_URL = "redis://localhost:6379"


class EvidenceSink(Protocol):
    """What the pipeline needs. A test substitutes a list; production substitutes Valkey."""

    def publish(self, record: dict) -> None: ...

    def close(self) -> None: ...


def to_record(shot: object) -> dict:
    """One `BestShot` as the wire record.

    Kept as a free function rather than a method so `attributes.py` stays pure CV with no knowledge
    of a broker, and so the payload shape is readable in one place next to the contract above.
    """
    from .attributes import BestShot  # noqa: PLC0415 — avoids a cycle; attributes stays I/O-free

    assert isinstance(shot, BestShot)
    return {
        "cameraId": shot.camera_id,
        "trackId": shot.track_id,
        "ts": shot.ts,
        "framePtsMs": shot.frame_pts_ms,
        "kind": "vehicle",
        "class": shot.vehicle_class,
        "detConfidence": round(float(shot.det_confidence), 3),
        "bbox": shot.bbox,
        "bestShotScore": shot.score,
        "focus": round(float(shot.focus), 2),
        "observations": shot.observations,
        "vehicleType": shot.body,
        # `unknown` when the read was refused — never the runner-up quietly promoted. The consumer
        # writes the name through as-is, so a low-confidence row is visible in SQL rather than
        # indistinguishable from a confident one.
        "vehicleColor": shot.color.name,
        "vehicleColorConfidence": shot.color.confidence,
        "attributesLowConfidence": shot.color.low_confidence,
        "colorChromaShare": shot.color.chroma_share,
        "colorRunnerUp": shot.color.runner_up,
        "contentType": "image/jpeg",
        # Base64 rather than a binary stream field: both clients (redis-py with
        # `decode_responses=True`, ioredis with its default string replies) round-trip text safely
        # and neither round-trips raw bytes without a second code path. The 33% is paid once per
        # track session, not once per sighting.
        "cropBase64": base64.b64encode(shot.crop_jpeg).decode("ascii"),
        "cropBytes": len(shot.crop_jpeg),
    }


@dataclass
class CollectingEvidenceSink:
    """Keeps everything, for tests and for the offline measurement runs."""

    records: list[dict] | None = None

    def __post_init__(self) -> None:
        if self.records is None:
            self.records = []

    def publish(self, record: dict) -> None:
        assert self.records is not None
        self.records.append(record)

    def close(self) -> None:
        return None


@dataclass
class NullEvidenceSink:
    """Counts and discards. Used by `bench.py`, which measures throughput, not persistence."""

    published: int = 0

    def publish(self, record: dict) -> None:
        self.published += 1

    def close(self) -> None:
        return None


def valkey_url() -> str:
    return os.environ.get("VALKEY_URL", DEFAULT_VALKEY_URL)


class ValkeyEvidenceSink:
    """`XADD` to the `evidence` stream.

    Failures are logged and counted, never raised — the same rule the sightings sink follows, for
    the same reason: a broker blip must not kill eight decode threads. An evidence record lost this
    way costs one crop, and the sighting row it belongs to is already safe on the other stream.
    """

    def __init__(self, url: str | None = None, maxlen: int = STREAM_MAXLEN) -> None:
        import redis  # noqa: PLC0415 — deferred; `valkey` speaks the redis protocol

        self._client = redis.Redis.from_url(url or valkey_url(), decode_responses=True)
        self._maxlen = maxlen
        self.published = 0
        self.failed = 0
        self.bytes_published = 0

    def publish(self, record: dict) -> None:
        try:
            self._client.xadd(
                STREAM_KEY,
                {"payload": json.dumps(record, separators=(",", ":"))},
                maxlen=self._maxlen,
                approximate=True,
            )
            self.published += 1
            self.bytes_published += int(record.get("cropBytes", 0))
        except Exception as exc:  # noqa: BLE001 — a broker blip must not kill a decode thread
            self.failed += 1
            if self.failed <= 5:
                log.warning("XADD to %s failed (%s): %s", STREAM_KEY, type(exc).__name__, exc)

    def close(self) -> None:
        try:
            self._client.close()
        except Exception as exc:  # noqa: BLE001
            log.debug("valkey close: %s", exc)
